stellar-multitenant/pkg/internal/provisioning/manager.go

```diff
@@ -49,238 +59,261 @@ func rollbackTenantDataSetupErrors() []error {
-		t, err = m.tenantManager.UpdateTenantConfig(
-			ctx,
-			&tenant.TenantUpdate{
-				ID:                  t.ID,
-				Status:              &tenantStatus,
-				SDPUIBaseURL:        &uiBaseURL,
-				DistributionAccount: t.DistributionAccount,
-			})
-		if err != nil {
-			updateTenantErrMsg := fmt.Errorf("%w: updating tenant %s status to %s: %w", ErrUpdateTenantFailed, name, tenant.ProvisionedTenantStatus, err)
-			return t, updateTenantErrMsg
-		}
+	provisioningErr := fmt.Errorf("provisioning error: %w", err)
 
-		return t, nil
-	}()
-	if err != nil {
-		// We don't want to roll back an existing tenant
-		if errors.Is(err, tenant.ErrDuplicatedTenantName) {
-			return nil, err
+	if errors.Is(err, ErrUpdateTenantFailed) {
+		log.Ctx(ctx).Errorf("tenant record not updated")
+	}
+
+	if isErrorInArray(err, deleteDistributionKeyErrors()) {
+		deleteDistributionAccFromVaultErr := m.deleteDistributionAccountKey(ctx, t)
+		// We should not let any failures from key deletion block us from completing the tenant cleanup process
+		if deleteDistributionAccFromVaultErr != nil {
+			deleteDistributionKeyErrPrefixMsg := fmt.Sprintf("deleting distribution account private key %s", *t.DistributionAccount)
+			provisioningErr = fmt.Errorf("%w. [additional errors]: %s: %w", provisioningErr, deleteDistributionKeyErrPrefixMsg, deleteDistributionAccFromVaultErr)
+			log.Ctx(ctx).Errorf("%s: %v", deleteDistributionKeyErrPrefixMsg, deleteDistributionAccFromVaultErr)
 		}
+		log.Ctx(ctx).Errorf("distribution account cleanup successful")
+	}
 
-		provisioningErr := fmt.Errorf("provisioning error: %w", err)
+	if isErrorInArray(err, rollbackTenantDataSetupErrors()) {
+		log.Ctx(ctx).Errorf("tenant data setup requires rollback")
+	}
 
-		if errors.Is(err, ErrUpdateTenantFailed) {
-			log.Ctx(ctx).Errorf("tenant record not updated")
+	if isErrorInArray(err, rollbackTenantCreationErrors()) {
+		deleteTenantErr := m.tenantManager.DeleteTenantByName(ctx, name)
+		if deleteTenantErr != nil {
+			return fmt.Errorf("%w. [rollback error] %w", provisioningErr, deleteTenantErr)
 		}
 
-		for _, deleteDistributionKeyErr := range deleteDistributionKeyErrors() {
-			if errors.Is(err, deleteDistributionKeyErr) {
-				deleteDistributionAccFromVaultErr := m.deleteDistributionAccountKey(ctx, t)
-				// We should not let any failures from key deletion block us from completing the tenant cleanup process
-				if deleteDistributionKeyErr != nil {
-					deleteDistributionKeyErrPrefixMsg := fmt.Sprintf("deleting distribution account private key %s", *t.DistributionAccount)
-					provisioningErr = fmt.Errorf("%w. [additional errors]: %s: %w", provisioningErr, deleteDistributionKeyErrPrefixMsg, deleteDistributionAccFromVaultErr)
-					log.Ctx(ctx).Errorf("%s: %v", deleteDistributionKeyErrPrefixMsg, deleteDistributionAccFromVaultErr)
-				}
-
-				log.Ctx(ctx).Errorf("distribution account cleanup successful")
-				break
-			}
-		}
+		log.Ctx(ctx).Errorf("tenant %s deleted", name)
+	}
```

Comment:
I think we can merge `rollbackTenantCreationErrors()` and `rollbackTenantSchemaErrors()` because they both cover the same set of errors. 

Because they both cover the same errors 
```go 
func rollbackTenantCreationErrors() []error {
	return []error{ErrUpdateTenantFailed, ErrProvisionTenantDistributionAccountFailed, ErrTenantDataSetupFailed, ErrTenantSchemaFailed}
}

func rollbackTenantSchemaErrors() []error {
	return []error{ErrUpdateTenantFailed, ErrProvisionTenantDistributionAccountFailed, ErrTenantDataSetupFailed, ErrTenantSchemaFailed}
}
```